# Analyzing High-Speed Railway Delays Project

- [View Solution Notebook](./solution.html)
- [View Project Page](https://www.codecademy.com/projects/practice/analyzing-high-speed-railways-with-python)

## Task Group 1 - Load and Clean

In [1]:
# Setup
import pandas as pd

df0 = pd.read_csv('datasets/railway_delays_2020_01_1.csv')
df0.head()

date           ride_id train_number  station_order  \
0  2020-01-01  G1226_20200101_1        G1226            1.0   
1  2020-01-01  G1226_20200101_2        G1226            2.0   
2  2020-01-01  G1226_20200101_3        G1226            3.0   
3  2020-01-01  G1226_20200101_4        G1226            4.0   
4  2020-01-01  G1226_20200101_5        G1226            5.0   

                  station_name                  to_station  mileage  \
0  Shenyangbei Railway Station    Shenyang Railway Station      5.0   
1     Shenyang Railway Station    Liaoyang Railway Station     65.0   
2     Liaoyang Railway Station    Anshanxi Railway Station     30.0   
3     Anshanxi Railway Station  Haichengxi Railway Station     39.0   
4   Haichengxi Railway Station      Panjin Railway Station     66.0   

   arrival_delay  departure_delay  major_holiday  
0            0.0             -1.0           True  
1           -2.0              2.0           True  
2            4.0              4.0           True  
3            5.0              4.0           True  
4            5.0              4.0           True

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What is the structure of this dataset? Toggle to check!</i></summary>

Each row contains departure and arrival delays for trains in China on January 1st, 2020. 

Here's a quick summary of the columns:

- **date** - the date of the train ride
- **ride_id** - the unique identifier for each train ride: `<train_number>_<yyyymmdd>_<station_order>`
- **train_number** - a unique identifier of the train
- **station_order** - the order in which each station was visited in a train route by a single train number on a specific date
- **station_name** - the name of the departing station
- **to_station** - the name of the next station destination
- **mileage** - the distance in kilometers to the next station
- **arrival_delay** - the delay in minutes arriving to the next station
- **departure_delay** - the delay in minutes departing the current station
- **major_holiday** - `True` if the `date` is a major holiday

</details>

### Task 1

Each day in January has its own data file. The names of these files depend on the date. For example, the file we read in above is `datasets/railway_delays_2020_01_1.csv`, corresponding to the day `2020-01-1`, or January 1st.

Use a list comprehension to create a list of the filenames from January 1st, 2020 to January 27th, 2020. Name this list `filenames` and print out the list. 

In [2]:
filenames = ['datasets/railway_delays_2020_01_' + str(i) + '.csv' for i in range(1,28)]
filenames

['datasets/railway_delays_2020_01_1.csv',
 'datasets/railway_delays_2020_01_2.csv',
 'datasets/railway_delays_2020_01_3.csv',
 'datasets/railway_delays_2020_01_4.csv',
 'datasets/railway_delays_2020_01_5.csv',
 'datasets/railway_delays_2020_01_6.csv',
 'datasets/railway_delays_2020_01_7.csv',
 'datasets/railway_delays_2020_01_8.csv',
 'datasets/railway_delays_2020_01_9.csv',
 'datasets/railway_delays_2020_01_10.csv',
 'datasets/railway_delays_2020_01_11.csv',
 'datasets/railway_delays_2020_01_12.csv',
 'datasets/railway_delays_2020_01_13.csv',
 'datasets/railway_delays_2020_01_14.csv',
 'datasets/railway_delays_2020_01_15.csv',
 'datasets/railway_delays_2020_01_16.csv',
 'datasets/railway_delays_2020_01_17.csv',
 'datasets/railway_delays_2020_01_18.csv',
 'datasets/railway_delays_2020_01_19.csv',
 'datasets/railway_delays_2020_01_20.csv',
 'datasets/railway_delays_2020_01_21.csv',
 'datasets/railway_delays_2020_01_22.csv',
 'datasets/railway_delays_2020_01_23.csv',
 'datasets/railway_d

### Task 2

Now that we have a list of filenames, we can import them as DataFrames. Use a list comprehension to create a list of DataFrames named `railway_delays_dfs`. Print summary information about the first DataFrame and take a look at the column names and their datatypes.

In [9]:
railway_delays_df = [pd.read_csv(x) for x in filenames]
railway_delays_df

[            date           ride_id train_number  station_order  \
 0     2020-01-01  G1226_20200101_1        G1226            1.0   
 1     2020-01-01  G1226_20200101_2        G1226            2.0   
 2     2020-01-01  G1226_20200101_3        G1226            3.0   
 3     2020-01-01  G1226_20200101_4        G1226            4.0   
 4     2020-01-01  G1226_20200101_5        G1226            5.0   
 ...          ...               ...          ...            ...   
 1106  2020-01-01  G868_20200101_24         G868           24.0   
 1107  2020-01-01  G868_20200101_25         G868           25.0   
 1108  2020-01-01  G868_20200101_26         G868           26.0   
 1109  2020-01-01  G868_20200101_27         G868           27.0   
 1110  2020-01-01  G868_20200101_28         G868           28.0   
 
                         station_name                      to_station  mileage  \
 0        Shenyangbei Railway Station        Shenyang Railway Station      5.0   
 1           Shenyang Railway 

### Task 3

Concatenate the DataFrames in `railway_delays_dfs` into a single DataFrame named `railway_delays_full`. Print summary information about `railway_delays_full`.

What do you notice about the columns in our fully concatenated DataFrame?

In [10]:
railway_delays_full = pd.concat(railway_delays_df)
railway_delays_full.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 29717 entries, 0 to 1081
Data columns (total 21 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   date               28604 non-null  object 
 1   ride_id            28604 non-null  object 
 2   train_number       28604 non-null  object 
 3   station_order      28604 non-null  float64
 4   station_name       27491 non-null  object 
 5   to_station         28604 non-null  object 
 6   mileage            28604 non-null  float64
 7   arrival_delay      28604 non-null  float64
 8   departure_delay    28604 non-null  float64
 9   major_holiday      28604 non-null  object 
 10  DATE               1113 non-null   object 
 11  RIDE_ID            1113 non-null   object 
 12  TRAIN_NUMBER       1113 non-null   object 
 13  STATION_ORDER      1113 non-null   float64
 14  STATION_NAME       1113 non-null   object 
 15  TO_STATION         1113 non-null   object 
 16  MILEAGE            1113

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What happened when we concatenated all of the DataFrames? Toggle to check!</i></summary>

Vertically concatenating all of the DataFrames in `railway_delays_dfs` resulted in duplicate columns. The `.info()` output tells us that each column name appears at least twice. One set of names is lowercased and the other is uppercased. We also see the final column `station_name` is another duplicate that probably resulted from the presence of extra whitespaces in its name.

These discrepancies are most likely due to individual DataFrames containing a different set of column names. Since Python and pandas are case-sensitive, we'll need to standardize the column names in every DataFrame before concatenation. 
</details>

### Task 4

Standardize the column names in each of the railway delay DataFrames by 

- *lowercasing* each column name
- *removing* any extra whitespaces

Use a `for` loop to automate this process.

After standardizing the names, re-concatenate the cleaned DataFrames into a single DataFrame named `railway_delays_full` and preview the first five rows.

In [5]:
new_col_names = []
for col in railway_delays_full.columns:
    new_col_names.append(col.strip().lower())
railway_delays_full.columns = new_col_names
railway_delays_full

date            ride_id train_number station_order  \
0     2020-01-01   G1226_20200101_1        G1226           1.0   
1     2020-01-01   G1226_20200101_2        G1226           2.0   
2     2020-01-01   G1226_20200101_3        G1226           3.0   
3     2020-01-01   G1226_20200101_4        G1226           4.0   
4     2020-01-01   G1226_20200101_5        G1226           5.0   
...          ...                ...          ...           ...   
1077  2020-01-27  G7592_20200127_20        G7592          20.0   
1078  2020-01-27  G7592_20200127_21        G7592          21.0   
1079  2020-01-27  G7592_20200127_22        G7592          22.0   
1080  2020-01-27  G7592_20200127_23        G7592          23.0   
1081  2020-01-27  G7592_20200127_24        G7592          24.0   

                      station_name                    to_station mileage  \
0      Shenyangbei Railway Station      Shenyang Railway Station     5.0   
1         Shenyang Railway Station      Liaoyang Railway Station    65.0   
2         Liaoyang Railway Station      Anshanxi Railway Station    30.0   
3         Anshanxi Railway Station    Haichengxi Railway Station    39.0   
4       Haichengxi Railway Station        Panjin Railway Station    66.0   
...                            ...                           ...     ...   
1077    Nanjingnan Railway Station   Jiangningxi Railway Station    10.0   
1078   Jiangningxi Railway Station  Maanshandong Railway Station    32.0   
1079  Maanshandong Railway Station          Wuhu Railway Station    43.0   
1080          Wuhu Railway Station    Fanchangxi Railway Station    41.0   
1081    Fanchangxi Railway Station      Tongling Railway Station    35.0   

      arrival_delay  departure_delay  major_holiday  ... ride_id train_number  \
0               0.0             -1.0           True  ...     NaN          NaN   
1              -2.0              2.0           True  ...     NaN          NaN   
2               4.0              4.0           True  ...     NaN          NaN   
3               5.0              4.0           True  ...     NaN          NaN   
4               5.0              4.0           True  ...     NaN          NaN   
...             ...              ...            ...  ...     ...          ...   
1077            0.0              2.0          False  ...     NaN          NaN   
1078            2.0              1.0          False  ...     NaN          NaN   
1079            0.0             -1.0          False  ...     NaN          NaN   
1080           -1.0             -1.0          False  ...     NaN          NaN   
1081           -2.0             -1.0          False  ...     NaN          NaN   

     station_order station_name to_station mileage  arrival_delay  \
0              NaN          NaN        NaN     NaN            NaN   
1              NaN          NaN        NaN     NaN            NaN   
2              NaN          NaN        NaN     NaN            NaN   
3              NaN          NaN        NaN     NaN            NaN   
4              NaN          NaN        NaN     NaN            NaN   
...            ...          ...        ...     ...            ...   
1077           NaN          NaN        NaN     NaN            NaN   
1078           NaN          NaN        NaN     NaN            NaN   
1079           NaN          NaN        NaN     NaN            NaN   
1080           NaN          NaN        NaN     NaN            NaN   
1081           NaN          NaN        NaN     NaN            NaN   

      departure_delay  major_holiday station_name  
0                 NaN            NaN          NaN  
1                 NaN            NaN          NaN  
2                 NaN            NaN          NaN  
3                 NaN            NaN          NaN  
4                 NaN            NaN          NaN  
...               ...            ...          ...  
1077              NaN            NaN          NaN  
1078              NaN            NaN          NaN  
1079              NaN            NaN

### Task 5
Let's continue cleaning the DataFrame. Use a for loop and `if/elif` statements to

- *uppercase* all the object datatype columns
- *strip* any hidden extra whitespace from object columns
- *convert* all float datatype columns to integers

Preview the cleaned DataFrame to confirm the transformations.

In [13]:
for col in railway_delays_full.columns:
    if railway_delays_full[col].dtypes is 'object':
        railway_delays_full[col] = railway_delays_full[col].str.strip().upper()
    elif railway_delays_full[col].dtypes is 'float64':
        railway_delays_full[col] = pd.to_numeric(railway_delays_full[col])
        
railway_delays_full.dtypes

date                  object
ride_id               object
train_number          object
station_order        float64
station_name          object
to_station            object
mileage              float64
arrival_delay        float64
departure_delay      float64
major_holiday         object
DATE                  object
RIDE_ID               object
TRAIN_NUMBER          object
STATION_ORDER        float64
STATION_NAME          object
TO_STATION            object
MILEAGE              float64
ARRIVAL_DELAY        float64
DEPARTURE_DELAY      float64
MAJOR_HOLIDAY         object
station_name          object
dtype: object

## Task Group 2 - Impact of Holiday Travel

### Task 6

Let's see if major holidays had an impact on railway delay times.

Let's first see how many train rides actually occurred on major holidays compared to those that didn't. Use the `.value_counts()` method on the `major_holiday` column in `railway_delays_full`.

There are four major holidays in our dataset. How does the number of trains on holidays compare to the number of trains on non-holidays?

In [15]:
railway_delays_full.dtypes
railway_delays_full['major_holiday'].value_counts()

False    24330
True      4274
Name: major_holiday, dtype: int64

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What did we discover in Task 6? Toggle to check!</i></summary>

There are `4274` individual station->station trips on the major holidays, averaging to `1068` per day.
    
On the `23` non-holiday days in our dataset, there are `25443` station->station trips, averaging to `1106` per day.
    
Based on this rough back-of-the-envelope calculation, there do seem to be slightly fewer individual trips on the holidays. But we'd have to look at data for more months or talk to a subject-matter expert to know for sure.
</details>

### Task 7

Groupby `major_holiday` and calculate the average of `arrival_delay` and `departure_delay`.

In [19]:
holiday_delays = railway_delays_full.groupby('major_holiday').agg({'arrival_delay': 'mean',
                                                                  'departure_delay': 'mean'})
holiday_delays

arrival_delay  departure_delay
major_holiday                                
False              23.390834        22.385368
True               23.815863        22.924193

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What did we discover about the impact of holiday dates? Toggle to check!</i></summary>

There doesn't seem to be a major difference in average delays, though both are slightly higher on holidays than non-holidays (by less than a minute). We'd have to dig into this further to know if this is a fluke of the dates we happen to be looking at or an actual statistical impact. Either way, it probably isn't enough to make a noticeable difference to riders!

</details>

## Task Group 3 - Impact of Distance Between Stations

### Task 8
The distance between stations could impact delays. For example, having further to travel might correspond to increased delays.

In the next tasks, we'll investigate the distances between stations to see if there is an impact on train delays. 

Compute descriptive summary statistics on the distances from station to station by calling the `.describe()` method on the `mileage` column in `railway_delays_full`.

What is the mean mileage (kilometers)?

In [20]:
railway_delays_full['mileage'].describe()

count    28604.000000
mean        88.000734
std         79.547856
min          1.000000
25%         43.000000
50%         70.000000
75%        120.000000
max       1410.000000
Name: mileage, dtype: float64

### Task 9

Let's see if longer-than-average distances correspond to longer delays.

Create a column `long_distance` that is `True` in any row with `mileage` greater than `88` (the average).

In [21]:
railway_delays_full['long_distance'] = railway_delays_full['mileage'] > 88
railway_delays_full.head()

date           ride_id train_number  station_order  \
0  2020-01-01  G1226_20200101_1        G1226            1.0   
1  2020-01-01  G1226_20200101_2        G1226            2.0   
2  2020-01-01  G1226_20200101_3        G1226            3.0   
3  2020-01-01  G1226_20200101_4        G1226            4.0   
4  2020-01-01  G1226_20200101_5        G1226            5.0   

                  station_name                  to_station  mileage  \
0  Shenyangbei Railway Station    Shenyang Railway Station      5.0   
1     Shenyang Railway Station    Liaoyang Railway Station     65.0   
2     Liaoyang Railway Station    Anshanxi Railway Station     30.0   
3     Anshanxi Railway Station  Haichengxi Railway Station     39.0   
4   Haichengxi Railway Station      Panjin Railway Station     66.0   

   arrival_delay  departure_delay major_holiday  ... TRAIN_NUMBER  \
0            0.0             -1.0          True  ...          NaN   
1           -2.0              2.0          True  ...          NaN   
2            4.0              4.0          True  ...          NaN   
3            5.0              4.0          True  ...          NaN   
4            5.0              4.0          True  ...          NaN   

  STATION_ORDER STATION_NAME  TO_STATION MILEAGE ARRIVAL_DELAY  \
0           NaN          NaN         NaN     NaN           NaN   
1           NaN          NaN         NaN     NaN           NaN   
2           NaN          NaN         NaN     NaN           NaN   
3           NaN          NaN         NaN     NaN           NaN   
4           NaN          NaN         NaN     NaN           NaN   

   DEPARTURE_DELAY  MAJOR_HOLIDAY  station_name      long_distance  
0              NaN            NaN                NaN         False  
1              NaN            NaN                NaN         False  
2              NaN            NaN                NaN         False  
3              NaN            NaN                NaN         False  
4              NaN            NaN                NaN         False  

[5 rows x 22 columns]

### Task 10

Groupby `long_distance` and calculate the average `arrival_delay` and `departure_delay`.

In [24]:
distance_group = railway_delays_full.groupby('long_distance').agg({'arrival_delay': 'mean',
                                                                   'departure_delay': 'mean'})
distance_group

arrival_delay  departure_delay
long_distance                                
False              16.598184        15.140762
True               34.636589        34.412987

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What did we discover about the impact of distances between stations? Toggle to check!</i></summary>

Trips with longer-than-average distance have almost double the mean delays of shorter-than-average trips. While we can't be 100% certain that the distances from station to station are a contributing factor to delay times, this is promising evidence to investigate this phenomenon further. 

</details>

## Task Group 4 - Impact of Weather

### Task 11

Lastly, let's take a look at delays during ideal weather conditions and compare them to the delays during severe weather conditions.

Import the CSV file `datasets/weather.csv` into a DataFrame named `weather` and preview the first five rows.

In [26]:
weather = pd.read_csv('datasets/weather.csv')
weather.head()

ride_id                      wind weather  temperature
0  G1226_20200101_1    light winds from the N   sunny         -7.0
1  G1226_20200101_2    light winds from the N   sunny         -7.0
2  G1226_20200101_3    light winds from the N   sunny         -3.0
3  G1226_20200101_4  gentle breeze from the N   sunny         -3.0
4  G1226_20200101_5  gentle breeze from the N   sunny         -3.0

### Task 12

Let's clean `weather` by 
- *uppercasing* all of the text/object datatype columns
- *stripping* whitespace from the object columns
- *converting* all the float datatype columns to integers (we've checked that they should all be integers!) 

Use `for` loops and `if/elif` statements to automate this process.

Preview the modified DataFrame. 

In [27]:
for column in weather.columns:
    if weather[column].dtype == 'object':
        weather[column] = weather[column].str.upper()
        weather[column] = weather[column].str.strip()
    
    elif weather[column].dtype == 'float64':
        weather[column] = weather[column].astype('int64')

weather.head()

ride_id                      wind weather  temperature
0  G1226_20200101_1    LIGHT WINDS FROM THE N   SUNNY           -7
1  G1226_20200101_2    LIGHT WINDS FROM THE N   SUNNY           -7
2  G1226_20200101_3    LIGHT WINDS FROM THE N   SUNNY           -3
3  G1226_20200101_4  GENTLE BREEZE FROM THE N   SUNNY           -3
4  G1226_20200101_5  GENTLE BREEZE FROM THE N   SUNNY           -3

### Task 13

Merge `weather` with `railway_delays_full` using the `ride_id` column. Use an inner merge and name the merged DataFrame `railway_delays_join_weather`. Preview the first five rows of the DataFrame.

In [28]:
railway_delays_join_weather = pd.merge(railway_delays_full, weather, on='ride_id', how='inner')
railway_delays_join_weather.head()

date           ride_id train_number  station_order  \
0  2020-01-01  G1226_20200101_1        G1226            1.0   
1  2020-01-01  G1226_20200101_2        G1226            2.0   
2  2020-01-01  G1226_20200101_3        G1226            3.0   
3  2020-01-01  G1226_20200101_4        G1226            4.0   
4  2020-01-01  G1226_20200101_5        G1226            5.0   

                  station_name                  to_station  mileage  \
0  Shenyangbei Railway Station    Shenyang Railway Station      5.0   
1     Shenyang Railway Station    Liaoyang Railway Station     65.0   
2     Liaoyang Railway Station    Anshanxi Railway Station     30.0   
3     Anshanxi Railway Station  Haichengxi Railway Station     39.0   
4   Haichengxi Railway Station      Panjin Railway Station     66.0   

   arrival_delay  departure_delay major_holiday  ... TO_STATION MILEAGE  \
0            0.0             -1.0          True  ...        NaN     NaN   
1           -2.0              2.0          True  ...        NaN     NaN   
2            4.0              4.0          True  ...        NaN     NaN   
3            5.0              4.0          True  ...        NaN     NaN   
4            5.0              4.0          True  ...        NaN     NaN   

  ARRIVAL_DELAY  DEPARTURE_DELAY MAJOR_HOLIDAY station_name       \
0           NaN              NaN           NaN               NaN   
1           NaN              NaN           NaN               NaN   
2           NaN              NaN           NaN               NaN   
3           NaN              NaN           NaN               NaN   
4           NaN              NaN           NaN               NaN   

   long_distance                      wind  weather temperature  
0          False    LIGHT WINDS FROM THE N    SUNNY          -7  
1          False    LIGHT WINDS FROM THE N    SUNNY          -7  
2          False    LIGHT WINDS FROM THE N    SUNNY          -3  
3          False  GENTLE BREEZE FROM THE N    SUNNY          -3  
4          False  GENTLE BREEZE FROM THE N    SUNNY          -3  

[5 rows x 25 columns]

### Task 14

Groupby the `weather` column, and calculate the average of `arrival_delay` and `departure_delay`. Sort the result from highest `arrival_delay` to lowest.

In [31]:
weather_grouped = railway_delays_join_weather.groupby('weather').agg({'arrival_delay': 'mean',
                                                                      'departure_delay': 'mean'})
weather_grouped

arrival_delay  departure_delay
weather                                               
BLIZZARD                   314.666667       314.666667
CLOUDY                      28.245066        27.157444
DOWNPOUR                     2.571429         1.771429
FOG                          0.701613         0.048387
HAZE                         0.501946         0.163424
HEAVY HAZE                  -0.125000        -0.166667
HEAVY SNOW                  79.289157        78.590361
LIGHT RAIN                  14.886330        13.686517
LIGHT SNOW                  51.291667        50.772569
LIGHT TO MODERATE RAIN       3.455197         1.483871
MODERATE HAZE                0.650943         0.490566
MODERATE RAIN               39.685804        38.501044
MODERATE SNOW               67.816794        66.977099
MODERATE TO HEAVY RAIN       0.833333         0.333333
OVERCAST                    19.017050        18.426418
SHOWERS                      7.551724         7.195402
SLEET                       39.614767        39.579454
SNOW SHOWERS                 4.328125         3.875000
SUNNY                       22.880296        21.694069

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What did we discover about the impact of weather? Toggle to check!</i></summary>

There is a noticeable difference in delays based on weather conditions.

- Severe weather conditions like blizzards and heavy snow saw a significant increase in delay times
- Ideal weather conditions like sunny and rainy conditions saw shorter delay times
- Fog and hazy weather conditions had the lowest delay times

The last bullet point is perhaps the most interesting. Can you think of any reasons why fog would have lower delays than sunny conditions?

</details>

There's definitely more insights to discover from this real-world dataset. Feel free to continue exploring here, or [download the source dataset](https://figshare.com/articles/dataset/A_high-speed_railway_network_dataset_from_train_operation_records_and_weather_data/15087882/4), which contains several more months of high-speed rail data!